Whylogs/whylabs integration

stefannica opened the pull request:
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Adds whylogs/whylabs as a ZenML integration:

* whylogs DatasetProfiles are treated as ZenML artifacts
* adds a materializer for DatasetProfiles. If whylabs credentials are set, the materializer will also upload the DatasetProfiles to whylabs, in addition to storing them in the artifact store
* adds a pipeline step visualizer for DatasetProfiles. It leverages the visualization support shipped with whylogs.
* adds a whylogs step context that acts as a session manager and wraps around the whylogs dataset profiling API
* adds a base whylogs profiling step as well as a step instantiation function that can be used to quickly create whylogs profiling steps and add them to a pipeline
* includes a BaseStep fix required to allow using StepContext subclasses in steps

Still missing, to be added:
- [x] example readme file
- [ ] example jupyter notebook
- [x] add example to tests

Optional improvements:
- [ ] group jupyter notebook diagrams in tabs

htahir1 commented on examples/whylogs/run.py:
```diff
@@ -0,0 +1,108 @@
+
+logger = get_logger(__name__)
+
+# Simply set these environment variables to point to a Whylabs account and all
```
This is already fantastic!

htahir1 commented on src/zenml/integrations/whylogs/materializers/__init__.py:
```diff
@@ -0,0 +1,16 @@
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+from zenml.integrations.whylogs.materializers.whylogs_materializer import (  # noqa
+    WhylogsMaterializer,
```
Is this a necessary import? The base __init__ already does it

bcdurak replied:
I believe this is useful for people who want to import this materializer directly from the integration.

bcdurak replied:
Also, the base init should only import it when the integration gets activated.

stefannica replied:
If by the base init, you mean `zenml.integrations.whylogs`, then no, I didn't add it there. But you're right, it's not really needed.

htahir1 replied:
Let's then leave it as it is, ill resolve all related comments

htahir1 commented on src/zenml/integrations/whylogs/materializers/whylogs_materializer.py:
```diff
@@ -0,0 +1,60 @@
+
+
+class WhylogsMaterializer(BaseMaterializer):
+    """Materializer to read/write Pytorch models."""
```
Typo: Whylogs not Pytorch

stefannica replied:
fixed

htahir1 commented on src/zenml/integrations/whylogs/materializers/whylogs_materializer.py:
```diff
@@ -0,0 +1,60 @@
+        with fileio.open(filepath, "wb") as f:
+            f.write(protobuf)
+
+        # TODO: find another way to detect if whylabs is enabled
```
wrong TODO format. Please do `TODO [LOW|MEDIUM|HIGH]: Description`

stefannica replied:
done

htahir1 commented on src/zenml/integrations/whylogs/steps/__init__.py:
```diff
@@ -0,0 +1,19 @@
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+from zenml.integrations.whylogs.steps.whylogs_profiler import (
```
Not sure if these are necessary either. Are you perhaps doing it to make the namespace easy to navigate?

stefannica replied:
I'm merely applying what I thought to be a convention used in many other integrations, one that I particularly like too, because instead of:

```from zenml.integrations.whylogs.steps.whylogs_profiler import WhylogsProfilerStep```

the end user can use the shorter form:

```from zenml.integrations.whylogs.steps import WhylogsProfilerStep```

htahir1 commented on src/zenml/integrations/whylogs/steps/whylogs_profiler.py:
```diff
@@ -0,0 +1,129 @@
+        tags: custom metadata tags associated with the whylogs profile
+
+    Also see `WhylogsContext.log_dataframe`.
+
```
Extra endline

htahir1 commented on src/zenml/integrations/whylogs/steps/whylogs_profiler.py:
```diff
@@ -0,0 +1,129 @@
+        tags: custom metadata tags associated with the whylogs profile
+    Returns:
+        a WhylogsProfilerStep step instance
+
```
extra endline

htahir1 commented on src/zenml/integrations/whylogs/visualizers/__init__.py:
```diff
@@ -0,0 +1,18 @@
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+from zenml.integrations.whylogs.visualizers.whylogs_visualizer import (  # noqa
```
Same as above two comments

stefannica replied:
Same as the other two: just used as a convenience

htahir1 commented on src/zenml/steps/utils.py:
```diff
@@ -393,7 +393,7 @@ def Do(
             elif issubclass(arg_type, StepContext):
                 output_artifacts = {k: v[0] for k, v in output_dict.items()}
-                context = StepContext(
+                context = arg_type(
```
For my information, why did this need to change?

stefannica replied:
This change is needed to cover the case where the step function signature specifies a subclass of `StepContext` instead, as is the case with `WhylogsStepContext`.

htahir1 replied:
Ah makes sense!

schustmi commented on src/zenml/integrations/whylogs/whylogs_utils.py:
```diff
@@ -0,0 +1,120 @@
+from zenml.steps.step_context import StepContext
+
+
+class WhylogsContext(StepContext):
```
The implementation as a separate context is really clean, however I do have one concern: What if the user requires multiple integrations (lets say whylogs and mlflow) in one step, would that require multiple contexts in the signature? (We currently only allow a single step context, but that is easily changed. I just fear that the step signature will get too messy)

bcdurak replied:
I share the same concern with @schustmi . Figuring out a way to be able to use multiple integrations with a single StepContext might be critical here.

htahir1 replied:
As a potential solution, we could append an object or a function to the generic `context` variable if certain integrations are activated, e.g.:

```python
context.whylogs  # monkeypatch
# or
context.get_integration('whylogs')
# or
context.get_whylogs()
```

htahir1 commented on src/zenml/integrations/whylogs/steps/whylogs_profiler.py:
```diff
@@ -0,0 +1,129 @@
+        return profile
+
+
+def whylogs_profiler_step(
```
So this utility is simply to create a new instance of the step to get around that error you ran into with duplicate steps?

stefannica replied:
Yes.

htahir1 commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,129 @@
+
+    @classmethod
+    def list(cls) -> List[str]:
+        names = map(lambda c: c.name, cls)  # type: ignore
```
ohh a `map` function!

htahir1 commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,129 @@
+
+
+class WhylogsVisualizer(BaseStepVisualizer):
+    """The implementation of an Whylogs Visualizer."""
```
`a` not `an` i believe

htahir1 commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,129 @@
+        profile: DatasetProfile,
+        plots: Optional[List[WhylogsPlots]] = None,
+    ) -> None:
+        """Generate a Facet Overview to visualize a whylogs dataset profile
```
A Facet Overview? Not sure if that word makes sense

schustmi commented on src/zenml/integrations/whylogs/whylogs_utils.py:
```diff
@@ -0,0 +1,120 @@
+
+        return self._session
+
+    def log_dataframe(
```
If I understood correctly then this method does not actually log/write/upload anything unless it is materialized, correct? If that's the case I feel like we should clarify that by either changing the method name or mentioning it in the docstring somewhere.

stefannica replied:
Correct ! And a very good point.

htahir1 commented on src/zenml/integrations/whylogs/whylogs_utils.py:
```diff
@@ -0,0 +1,120 @@
+
+        self._session = Session(
+            project=self.step_name,
+            # TODO: replace with pipeline name when available in the context
```
Wrong TODO format

htahir1 commented on src/zenml/integrations/whylogs/whylogs_utils.py:
```diff
@@ -0,0 +1,120 @@
+        # name across different pipelines. The pipeline name is also required
+        dataset_name = dataset_name or self.step_name
+        tags = tags or dict()
+        # TODO: add more tags when this information is available in the context:
```
Wrong TODO format

schustmi commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,129 @@
+                    plot_method = self._get_plot_method(visualizer, plot)
+                    display(plot_method(column))
+        else:
+            logger.warn(
```
Nit: I think this is deprecated and should be `logger.warning()` instead :P

htahir1 reviewed (changes requested):
Okay, this is excellent work! I ran it and it gave a really nice visualization. I think you thought of this deeply and the work must be appreciated. I just have a few minor comments here and there.

In general, I am unsure how I feel about overriding the Context variable to pass in the WhyLogsContext. I liked what @AlexejPenner implemented with mlflow, i.e., passing it in via a decorator and then importing it at the top. In this case it would look like

```
from zenml.integrations.whylogs import get_whylogs_session

@step
def my_step()
    session = get_whylogs_session()

@enable_whylogs
def pipeline()
    ... 
```

But i lack the nuanced understanding of whylogs to know whether that is better or worse. Happty to discuss it!

bcdurak reviewed (commented):
Aside from a few comments, it looks really great 👍 One possible addition could be a `README.md` file to the corresponding example.

stefannica commented:
> Okay, this is excellent work! I ran it and it gave a really nice visualization. I think you thought of this deeply and the work must be appreciated. I just have a few minor comments here and there.
> 
> In general, I am unsure how I feel about overriding the Context variable to pass in the WhyLogsContext. I liked what @AlexejPenner implemented with mlflow, i.e., passing it in via a decorator and then importing it at the top. In this case it would look like
> 
> ```
> from zenml.integrations.whylogs import get_whylogs_session
> 
> @step
> def my_step()
>     session = get_whylogs_session()
> 
> @enable_whylogs
> def pipeline()
>     ... 
> ```
> 
> But i lack the nuanced understanding of whylogs to know whether that is better or worse. Happty to discuss it!

I'm not too happy about it either. Using a derived context seemed like a simple way to stick the session constructor to something that could be easily accessed from within a step at runtime. Now I'm having second thoughts about it, especially given @schustmi 's observation about not scaling well when multiple integrations are stacked together. I'll try to find something else less intrusive.

stefannica commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,129 @@
+            # filter out anything but whylog dataset profile artifacts
+            if artifact_view.data_type == ".".join(
+                [DatasetProfile.__module__, DatasetProfile.__name__]
+            ):
```
Good idea !

stefannica commented:
@htahir1 @schustmi I addressed all your comments in this last commit, with the exception of the derived context.

stefannica commented:
@htahir1 @bcdurak @schustmi I reworked the whylogs context approach and I implemented an `enable_whylogs` step decorator that adds a `whylogs` field to the step context. This scales much better than sub-classing the StepContext, with the downside that it's a bit more complicated to implement.

stefannica commented:
> Aside from a few comments, it looks really great +1 One possible addition could be a `README.md` file to the corresponding example.

Thanks @bcdurak . The README is now included with the example.

htahir1 reviewed (approved):
LGTM! Awesome job, beautifully implemented!

schustmi reviewed (changes requested):
Really nice job! Just two tiny requests

jwwwb commented on examples/whylogs/run.py:
```diff
@@ -0,0 +1,129 @@
+    """Load the diabetes dataset."""
+    X, y = datasets.load_diabetes(return_X_y=True, as_frame=True)
+
+    # merge X an y together
```
This is a tiny thing, only if you have other commits to do, this should read "and".

jwwwb commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,122 @@
+class WhylogsPlots(StrEnum):
+    """All supported whylogs plot types."""
+
+    distribution = "plot_distribution"
```
Aren't enum values usually in yelling case?

jwwwb commented on src/zenml/utils/enum_utils.py:
```diff
@@ -0,0 +1,30 @@
+    @classmethod
+    def list(cls) -> List[str]:
+        """Get all enum names as a list of strings"""
+        names = map(lambda c: c.name, cls)  # type: ignore
```
I'm a little curious about this kind of roundabout approach of using a map, then casting to list using `typing` module and then casting to list with the python builtin? Would it not also work to just define this function as:
```return [c.name for c in cls]```
?

stefannica commented on src/zenml/integrations/whylogs/steps/whylogs_profiler.py:
```diff
@@ -0,0 +1,129 @@
+    whylogs DatasetProfile from a given pd.DataFrame and save it as an artifact.
+
+    Args:
+        name: The name of the step
```
Done

stefannica commented on src/zenml/integrations/whylogs/visualizers/whylogs_visualizer.py:
```diff
@@ -0,0 +1,122 @@
+    def _get_plot_method(
+        visualizer: ProfileVisualizer, plot: WhylogsPlots
+    ) -> Any:
+        plot_method = getattr(visualizer, plot, None)
```
Good idea, I included it
